File: refaker.py
import random
import argparse
import pyformance.reporters
import redis
from faker import Faker
from datetime import datetime, timedelta
from concurrent.futures import ProcessPoolExecutor
from concurrent.futures import ThreadPoolExecutor
import multiprocessing
import pyformance

def parse_args():
    parser = argparse.ArgumentParser(description="Read Redis connection arguments")
    parser.add_argument("-H", default="localhost", type=str, required=False, help="Redis host")
    parser.add_argument("-p", default=6379, type=int, required=False, help="Redis port")
    parser.add_argument("-a", default=None, type=str, required=False, help="Redis password")
    return parser.parse_args()

# Initialize Faker
fake = Faker()
r = []
# Connect to Redis (Assumes Redis is running on localhost:6379)
#r. = redis.Redis(host='localhost', port=6379, db=0)
registry = pyformance.MetricsRegistry()

# Function to generate a random timestamp within the last 2 years
def random_timestamp():
    today = datetime.now()
    start_date = today - timedelta(days=730)
    random_date = start_date + timedelta(days=random.randint(0, 730))
    return int(random_date.timestamp())

# Function to generate a single JSON object (person data)
def generate_person():
    return {
        "firstname": fake.first_name(),
        "lastname": fake.last_name(),
        "city": fake.city(),
        "country": fake.country(),
        "email": fake.email(),
        "age": random.randint(1, 99),
        "score": random.randint(0, 100),
        "friends": random.randint(0, 500),
        "registrationDate": random_timestamp(),
        "connectionDate": random_timestamp(),
    }

# Function to insert data into Redis using pipeline in batches
def insert_into_redis(threadID, start_index, count, batch_size=100):
    pipeline = r[threadID].pipeline()

    for i in range(count):
        person = generate_person()
        key = f"person:{start_index + i}"  # Generate a unique Redis key for each object
        pipeline.json().set(key, '$', person)  # Use RedisJSON to set the full JSON object
        
        # Execute pipeline when batch size is reached
        if (i + 1) % (batch_size) == 0:
            if (threadID==0):
                with registry.timer("pipeline").time():
                    pipeline.execute()
                #print(f"{start_index + i + 1} JSON objects inserted so far...", end='\r')
                print(f"{registry.timer('pipeline').get_mean()*1000/batch_size}",end='\r')
            else:
                pipeline.execute()
    # Execute the remaining objects in the pipeline if any
    if count % batch_size != 0:
        pipeline.execute()
        print(f"Final batch for thread starting at {start_index} inserted. Total {count} JSON objects inserted.")

# Main function to configure and run the insert process
def main():
    multiprocessing.set_start_method("fork")
    total_json = 2_000_000  # Set the total number of JSON objects to be created
    batch_size = 100   # Set the pipeline batch size
    num_threads = 20   # Set the number of concurrent threads
    # Calculate the number of objects each thread should handle
    count_per_thread = total_json // num_threads
    remainder = total_json % num_threads

    args = parse_args()

    for t in range(num_threads):
        r.append(redis.Redis(host=args.H, port=args.p, password=args.a, db=0))

    reporter = pyformance.reporters.ConsoleReporter(registry, reporting_interval=5)
    reporter.start()

    # Start threads to process data insertion concurrently
    while True:
        futures = []
        #with ThreadPoolExecutor(1) as executor:
        #    start_index = 0
        #    count = count_per_thread
        #    executor.submit(insert_into_redis, 0, start_index, count, batch_size)
        with ProcessPoolExecutor(max_workers=num_threads) as executor:
            for t in range(0,num_threads):
                start_index = t * count_per_thread
                count = count_per_thread + (1 if t < remainder else 0)  # Distribute any remainder
                futures.append(executor.submit(insert_into_redis, t, start_index, count, batch_size))

        # Wait for all threads to complete
        for future in futures:
            future.result()

        print("All data insertion tasks completed.")
        registry.clear()

if __name__ == "__main__":
    main()

File: refaker.go
package main

import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"math/rand"
	"sync"
	"time"

	"github.com/brianvoe/gofakeit/v6"
	"github.com/rcrowley/go-metrics"
	"github.com/redis/go-redis/v9"
)

type Person struct {
	Firstname        string `json:"firstname"`
	Lastname         string `json:"lastname"`
	City             string `json:"city"`
	Country          string `json:"country"`
	Email            string `json:"email"`
	Age              int    `json:"age"`
	Score            int    `json:"score"`
	Friends          int    `json:"friends"`
	RegistrationDate int64  `json:"registrationDate"`
	ConnectionDate   int64  `json:"connectionDate"`
}

var timer = metrics.NewTimer()

func randomTimestamp() int64 {
	start := time.Now().AddDate(-2, 0, 0).Unix()
	end := time.Now().Unix()
	return rand.Int63n(end-start) + start
}

func generatePerson() Person {
	return Person{
		Firstname:        gofakeit.FirstName(),
		Lastname:         gofakeit.LastName(),
		City:             gofakeit.City(),
		Country:          gofakeit.Country(),
		Email:            gofakeit.Email(),
		Age:              rand.Intn(98) + 1,
		Score:            rand.Intn(101),
		Friends:          rand.Intn(501),
		RegistrationDate: randomTimestamp(),
		ConnectionDate:   randomTimestamp(),
	}
}

func insertIntoRedis(ctx context.Context, client *redis.Client, startIndex, count, batchSize int, wg *sync.WaitGroup) {
	defer wg.Done()

	pipe := client.Pipeline()
	for i := 0; i < count; i++ {
		person := generatePerson()
		data, _ := json.Marshal(person)
		key := fmt.Sprintf("person:%d", startIndex+i)
		pipe.Do(ctx, "JSON.SET", key, "$", string(data))

		if (i+1)%batchSize == 0 {
			start := time.Now()
			_, err := pipe.Exec(ctx)
			timer.UpdateSince(start)

			if err != nil {
				log.Println("Pipeline error:", err)
			}
		}
	}
	// Final batch
	if count%batchSize != 0 {
		start := time.Now()
		_, err := pipe.Exec(ctx)
		timer.UpdateSince(start)

		if err != nil {
			log.Println("Final pipeline error:", err)
		}
	}
}

func reportMetrics(interval time.Duration, batchSize int, done <-chan struct{}) {
	ticker := time.NewTicker(interval)
	defer ticker.Stop()

	for {
		select {
		case <-done:
			return
		case <-ticker.C:
			snapshot := timer.Snapshot()
			fmt.Printf("[Metrics] Count: %d  Mean: %.2fms  95th: %.2fms  99th: %.2fms  Rate: %.2f/s\n",
				snapshot.Count(),
				float64(snapshot.Mean())/1e6,
				float64(snapshot.Percentile(0.95))/1e6,
				float64(snapshot.Percentile(0.99))/1e6,
				timer.Rate1()*float64(batchSize),
			)
			timer = metrics.NewTimer()
		}
	}
}

func main() {
	host := flag.String("h", "localhost", "Redis host")
	port := flag.Int("p", 6379, "Redis port")
	password := flag.String("a", "", "Redis password")
	flag.Parse()

	totalJSON := 2_000_000
	batchSize := 100
	numWorkers := 20
	countPerWorker := totalJSON / numWorkers
	remainder := totalJSON % numWorkers

	//rand.Seed(time.Now().UnixNano())
	gofakeit.Seed(time.Now().UnixNano())

	ctx := context.Background()
	done := make(chan struct{})
	go reportMetrics(5*time.Second, batchSize, done)

	for {
		var wg sync.WaitGroup
		for i := 0; i < numWorkers; i++ {
			startIndex := i * countPerWorker
			count := countPerWorker
			if i < remainder {
				count++
			}
			wg.Add(1)

			client := redis.NewClient(&redis.Options{
				Addr:     fmt.Sprintf("%s:%d", *host, *port),
				Password: *password,
				DB:       0,
			})

			go insertIntoRedis(ctx, client, startIndex, count, batchSize, &wg)
		}

		wg.Wait()
	}
}
